File: ios/WebyarNative/Sources/DesignSystem/Components/Controls.swift
import SwiftUI

// MARK: - Primary button

/// The one prominent action on a screen.
///
/// It is full width, `minTouchTarget` tall, and swaps its label for a spinner
/// while busy *without changing size* — a button that shrinks under the
/// thumb mid-tap is how a mis-tap happens.
struct PrimaryButton: View {
    let title: String
    var isLoading: Bool = false
    var isEnabled: Bool = true
    let action: () -> Void

    var body: some View {
        Button(action: action) {
            ZStack {
                // Keeps the intrinsic width while the spinner shows.
                Text(title)
                    .font(Theme.Typo.button)
                    .opacity(isLoading ? 0 : 1)

                if isLoading {
                    ProgressView()
                        .progressViewStyle(.circular)
                        .tint(.white)
                }
            }
            .frame(maxWidth: .infinity)
            .frame(height: Theme.Size.minTouchTarget + 6)
            .foregroundStyle(.white)
            .background {
                let shape = RoundedRectangle(cornerRadius: Theme.Radius.lg, style: .continuous)
                // A gradient rather than a flat fill, and only here: the one
                // committing action on a screen is the only thing in this app
                // allowed to carry the brand as a surface.
                shape.fill(Theme.Gradient.brand)
                    // Fading the whole button takes the white label down with
                    // it and leaves the title barely readable. Dimming only
                    // the fill keeps the text at full contrast, so a disabled
                    // button still says plainly what it will do.
                    .opacity(isEnabled && !isLoading ? 1 : 0.4)
                    .overlay {
                        // The lit top edge every raised iOS surface has. It
                        // is what stops a filled rectangle reading as a flat
                        // coloured box.
                        shape.strokeBorder(
                            LinearGradient(
                                colors: [.white.opacity(0.34), .white.opacity(0.04)],
                                startPoint: .top,
                                endPoint: .bottom
                            ),
                            lineWidth: 0.75
                        )
                    }
                    .elevated(isEnabled && !isLoading ? .raised : .resting)
            }
        }
        .disabled(!isEnabled || isLoading)
        .buttonStyle(PressableButtonStyle())
        .animation(Theme.Motion.standard, value: isLoading)
        .animation(Theme.Motion.standard, value: isEnabled)
    }
}

// MARK: - Segmented filter

/// The inbox queue filter.
///
/// `Picker(.segmented)` is the native control and already handles RTL, Dynamic
/// Type and the selection animation, so it is used rather than reimplemented.
/// The queues come from the plan, not from `allCases` — a segment that leads
/// to a permanently empty list because the plan excludes it reads as a broken
/// app, not as an upsell.
///
/// A count rides in the segment label when there is one, because the whole
/// reason to glance at this control is to see where the work is.
struct FilterPicker: View {

    /// A chip that leaves this screen rather than narrowing it.
    ///
    /// The strip is a filter control and these are not filters, so they are
    /// kept apart rather than mixed into `filters`: they sit after a divider,
    /// they carry an icon, and nothing can ever draw one as selected. An
    /// operator who taps one and comes back finds the queue they left still
    /// chosen, because it never stopped being chosen.
    struct Destination: Identifiable {
        let id: String
        let title: String
        /// SF Symbol, and the same one the title menu uses for this place.
        let icon: String
        let open: () -> Void
    }

    @Binding var selection: InboxFilter
    let filters: [InboxFilter]
    let counts: InboxCounts?
    let language: Language
    /// The inboxes that are their own screens. Empty unless the plan grants
    /// one, which is why this defaults rather than being passed everywhere.
    var destinations: [Destination] = []

    private func label(for filter: InboxFilter) -> String {
        let title = filter.chipTitle(language)
        guard let count = counts?.count(for: filter), count > 0 else { return title }
        return "\(title) (\(Format.number(count, language: language)))"
    }

    @Namespace private var chip
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    var body: some View {
        // Chips rather than a segmented control.
        //
        // `Picker(.segmented)` divides the width it is given equally, which
        // is fine for two short words and wrong for what this screen
        // actually shows: five queue names in Turkish, each with a count
        // riding in the label. Everything shrank to fit the longest one and
        // the counts -- the reason to glance at the control at all -- came
        // out at eight points. A scrolling row gives every chip the width of
        // its own words and the queue with forty waiting conversations can
        // say so.
        ScrollView(.horizontal) {
            LiquidGlassGroup(spacing: Theme.Space.lg) {
                HStack(spacing: Theme.Space.sm) {
                    ForEach(filters) { filter in
                        chipButton(for: filter)
                    }

                    if !destinations.isEmpty {
                        // A hairline, not a gap. The two groups are different
                        // kinds of thing — one narrows the list, one leaves
                        // it — and a little more space would only read as a
                        // layout accident.
                        Rectangle()
                            .fill(Theme.Palette.separator.opacity(0.7))
                            .frame(width: 0.5, height: 20)
                            .padding(.horizontal, Theme.Space.xxs)
                            .accessibilityHidden(true)

                        ForEach(destinations) { destination in
                            destinationButton(for: destination)
                        }
                    }
                }
                // Room for the glass edge and its shadow, which would
                // otherwise be shaved off by the scroll view's bounds.
                .padding(.vertical, Theme.Space.xs)
                .padding(.horizontal, Theme.Space.xxs)
            }
        }
        .scrollIndicators(.hidden)
        // The shadow under a chip belongs outside the scroll view's clip.
        .scrollClipDisabled()
        .accessibilityElement(children: .contain)
    }

    private func chipButton(for filter: InboxFilter) -> some View {
        let isSelected = filter == selection
        let count = counts?.count(for: filter) ?? 0

        return Button {
            guard !isSelected else { return }
            Haptics.selection()
            withAnimation(reduceMotion ? nil : Theme.Motion.morph) {
                selection = filter
            }
        } label: {
            HStack(spacing: Theme.Space.xs) {
                Text(filter.chipTitle(language))
                    .font(.app(.subheadline, weight: isSelected ? .semibold : .medium))
                    .lineLimit(1)

                if count > 0 {
                    Text(Format.number(count, language: language))
                        .font(Theme.Typo.metaEmphasis)
                        .monospacedDigit()
                        // The count changes under the operator's eyes as
                        // conversations arrive; rolling the digits says that
                        // plainly where a cross-fade just flickers.
                        .contentTransition(.numericText())
                        .padding(.horizontal, Theme.Space.xs)
                        .frame(minWidth: 20, minHeight: 19)
                        .background(
                            Capsule().fill(
                                isSelected
                                    ? Color.white.opacity(0.22)
                                    : Theme.Palette.brand.opacity(0.13)
                            )
                        )
                }
            }
            .foregroundStyle(isSelected ? Color.white : Theme.Palette.label)
            .padding(.horizontal, Theme.Space.sm + Theme.Space.xs)
            .frame(height: 36)
            .background {
                if isSelected {
                    Capsule(style: .continuous)
                        .fill(Theme.Gradient.brand)
                        .elevated(.resting)
                        .matchedGeometryEffect(id: "chip", in: chip)
                } else {
                    // A real surface, not glass.
                    //
                    // Glass is for chrome that content moves beneath -- the
                    // tab bar, a toolbar, the composer. These chips are a
                    // row *inside* the list, sitting still on an opaque page;
                    // a lens over an opaque page has nothing to refract and
                    // renders as a pale smear that its own label cannot be
                    // read on. White with a hairline is what an inline
                    // control on iOS actually looks like.
                    Capsule(style: .continuous)
                        .fill(Theme.Palette.surface)
                        .overlay(
                            Capsule(style: .continuous)
                                .strokeBorder(Theme.Palette.separator.opacity(0.7), lineWidth: 0.5)
                        )
                        .elevated(.resting)
                }
            }
            .contentShape(Capsule())
        }
        .buttonStyle(PressableButtonStyle())
        .accessibilityLabel(label(for: filter))
        .accessibilityAddTraits(isSelected ? [.isSelected, .isButton] : .isButton)
    }

    /// The same capsule as an unselected queue, with its icon in front of it.
    ///
    /// Deliberately never the selected treatment: this chip opens a screen
    /// and comes straight back off it, and a chip that lit up and then went
    /// out again as the operator returned would read as the tap having been
    /// undone.
    private func destinationButton(for destination: Destination) -> some View {
        Button {
            Haptics.selection()
            destination.open()
        } label: {
            HStack(spacing: Theme.Space.xs) {
                Image(systemName: destination.icon)
                    .font(.system(size: 12, weight: .semibold))
                    .foregroundStyle(Theme.Palette.brand)

                Text(destination.title)
                    .font(.app(.subheadline, weight: .medium))
                    .lineLimit(1)
            }
            .foregroundStyle(Theme.Palette.label)
            .padding(.horizontal, Theme.Space.sm + Theme.Space.xs)
            .frame(height: 36)
            .background {
                Capsule(style: .continuous)
                    .fill(Theme.Palette.surface)
                    .overlay(
                        Capsule(style: .continuous)
                            .strokeBorder(Theme.Palette.separator.opacity(0.7), lineWidth: 0.5)
                    )
                    .elevated(.resting)
            }
            .contentShape(Capsule())
        }
        .buttonStyle(PressableButtonStyle())
        .accessibilityLabel(destination.title)
        .accessibilityAddTraits(.isButton)
    }
}

// MARK: - Press feedback

/// A tap that the thumb can feel.
///
/// `.plain` leaves a custom-drawn button completely inert under the finger,
/// which on iOS reads as a tap that did not land -- the reason people tap a
/// second time. Every hand-drawn control in this app uses this instead, so
/// they all answer the same way.
struct PressableButtonStyle: ButtonStyle {
    @Environment(\.accessibilityReduceMotion) private var reduceMotion

    func makeBody(configuration: Configuration) -> some View {
        configuration.label
            .scaleEffect(configuration.isPressed && !reduceMotion ? 0.96 : 1)
            .opacity(configuration.isPressed ? 0.9 : 1)
            .animation(.spring(response: 0.22, dampingFraction: 0.72), value: configuration.isPressed)
    }
}

// MARK: - States

/// Shown when a list legitimately has nothing in it.
///
/// Centred as a block, with the icon, title and body on one vertical axis and
/// the body text centre-aligned and width-limited so it never runs edge to
/// edge.
struct EmptyStateView: View {
    let systemImage: String
    let title: String
    let message: String

    var body: some View {
        VStack(spacing: Theme.Space.md) {
            Image(systemName: systemImage)
                .font(.system(size: 34, weight: .light))
                .foregroundStyle(Theme.Palette.brand)
                .symbolRenderingMode(.hierarchical)
                .frame(width: 76, height: 76)
                // The glyph sits on its own disc rather than floating loose
                // in the middle of the screen. An empty state is still a
                // composed screen, not an absence of one.
                .liquidGlass(.card, in: Circle())

            VStack(spacing: Theme.Space.xs) {
                Text(title)
                    .font(.app(.headline))
                    .foregroundStyle(Theme.Palette.label)

                Text(message)
                    .font(.app(.subheadline))
                    .foregroundStyle(Theme.Palette.labelSecondary)
                    .multilineTextAlignment(.center)
                    // A measure this wide stays comfortable to read; full-width
                    // centred text does not.
                    .frame(maxWidth: 280)
            }
        }
        .frame(maxWidth: .infinity)
        .padding(.horizontal, Theme.screenInset)
        .padding(.vertical, Theme.Space.huge)
    }
}

/// Shown when a request failed and retrying is the sensible next step.
struct ErrorStateView: View {
    let title: String
    let message: String
    let retryTitle: String
    let onRetry: () -> Void

    var body: some View {
        VStack(spacing: Theme.Space.md) {
            Image(systemName: "wifi.exclamationmark")
                .font(.system(size: 40, weight: .light))
                .foregroundStyle(Theme.Palette.warning)

            VStack(spacing: Theme.Space.xs) {
                Text(title)
                    .font(.app(.headline))
                Text(message)
                    .font(.app(.subheadline))
                    .foregroundStyle(Theme.Palette.labelSecondary)
                    .multilineTextAlignment(.center)
                    .frame(maxWidth: 280)
            }

            Button(action: onRetry) {
                Text(retryTitle)
                    .font(Theme.Typo.button)
                    .padding(.horizontal, Theme.Space.xl)
                    .frame(height: Theme.Size.minTouchTarget)
            }
            .buttonStyle(.bordered)
            .buttonBorderShape(.capsule)
            .padding(.top, Theme.Space.xs)
        }
        .frame(maxWidth: .infinity)
        .padding(.horizontal, Theme.screenInset)
        .padding(.vertical, Theme.Space.huge)
    }
}

// MARK: - Badges

/// The unread counter on an inbox row.
struct UnreadBadge: View {
    let count: Int
    @Environment(\.locale) private var locale

    var body: some View {
        // Formatted rather than interpolated. A ternary between two string
        // literals settles on `String`, which is the overload of `Text` that
        // does *not* localize — so the capped form would have shipped as an
        // ASCII "99+" beside Persian digits on the row above it.
        Text(count > 99 ? "\(Format.number(99, locale: locale))+" : Format.number(count, locale: locale))
            .font(Theme.Typo.metaEmphasis)
            .monospacedDigit()
            .foregroundStyle(.white)
            .contentTransition(.numericText())
            .padding(.horizontal, Theme.Space.sm)
            .frame(minWidth: 22, minHeight: 20)
            .background(Capsule().fill(Theme.Gradient.brand).elevated(.resting))
            // A conversation just arrived in a queue the operator is looking
            // at. The digits already roll; this is the badge itself noticing.
            .jumpsOnChange(count)
            // The digits themselves localize with the reader's language; this
            // only stops the capped form rendering as "+99" under RTL.
            .environment(\.layoutDirection, .leftToRight)
    }
}

/// A small status pill — "resolved", "AI", and similar.
struct StatusPill: View {
    let text: String
    let tint: Color

    var body: some View {
        Text(text)
            .font(Theme.Typo.meta)
            .foregroundStyle(tint)
            .padding(.horizontal, Theme.Space.sm)
            .padding(.vertical, Theme.Space.xxs)
            .background {
                Capsule().fill(tint.opacity(0.13))
                    // A hairline of the same colour keeps the pill legible
                    // where the fill alone is nearly the page behind it.
                    .overlay(Capsule().strokeBorder(tint.opacity(0.22), lineWidth: 0.5))
            }
    }
}


/// Forces left-to-right on a run of text that is Latin whatever the interface
/// language is — an address, a phone number, a version string.
///
/// Reading `operator@webyar.app` right-to-left puts the domain first, which is
/// wrong in Persian and Turkish just as it would be in English.
struct LatinIfNeeded: ViewModifier {
    let isLatin: Bool

    func body(content: Content) -> some View {
        if isLatin {
            content.environment(\.layoutDirection, .leftToRight)
        } else {
            content
        }
    }
}

extension View {
    func latin(_ isLatin: Bool = true) -> some View {
        modifier(LatinIfNeeded(isLatin: isLatin))
    }
}

/// A centred, quiet line inside a list — "nothing here yet", said without
/// making a scene of it.
///
/// Not an `EmptyStateView`: that one owns a whole screen with an icon and a
/// title. This is one row saying one section is empty while the rest of the
/// list carries on around it.
struct QuietRow: View {
    let text: String

    var body: some View {
        Text(text)
            .font(.app(.subheadline))
            .foregroundStyle(Theme.Palette.labelSecondary)
            .frame(maxWidth: .infinity, alignment: .center)
            .padding(.vertical, Theme.Space.sm)
            .listRowSeparator(.hidden)
    }
}

File: android/app/src/main/kotlin/com/webyar/operator/ui/components/Controls.kt
package com.webyar.operator.ui.components

import androidx.compose.foundation.layout.Arrangement
import androidx.compose.foundation.layout.Box
import androidx.compose.foundation.layout.Column
import androidx.compose.foundation.layout.PaddingValues
import androidx.compose.foundation.layout.Row
import androidx.compose.foundation.layout.Spacer
import androidx.compose.foundation.layout.fillMaxWidth
import androidx.compose.foundation.layout.heightIn
import androidx.compose.foundation.layout.padding
import androidx.compose.foundation.layout.size
import androidx.compose.foundation.layout.widthIn
import androidx.compose.foundation.shape.RoundedCornerShape
import androidx.compose.material3.Button
import androidx.compose.material3.Icon
import androidx.compose.material3.MaterialTheme
import androidx.compose.material3.Surface
import androidx.compose.material3.Text
import androidx.compose.runtime.Composable
import androidx.compose.ui.Alignment
import androidx.compose.ui.Modifier
import androidx.compose.ui.platform.LocalLayoutDirection
import androidx.compose.ui.graphics.vector.ImageVector
import androidx.compose.ui.text.style.TextAlign
import androidx.compose.ui.text.style.TextDirection
import androidx.compose.ui.text.style.TextOverflow
import androidx.compose.ui.text.TextStyle
import androidx.compose.ui.unit.LayoutDirection
import androidx.compose.ui.unit.dp
import com.webyar.operator.i18n.Format
import com.webyar.operator.i18n.Language
import com.webyar.operator.ui.design.Radius
import com.webyar.operator.ui.design.Size
import com.webyar.operator.ui.design.Space
import com.webyar.operator.ui.design.WebyarTheme
import androidx.compose.animation.AnimatedVisibility
import androidx.compose.animation.animateColorAsState
import androidx.compose.animation.core.animateFloatAsState
import androidx.compose.foundation.horizontalScroll
import androidx.compose.foundation.interaction.MutableInteractionSource
import androidx.compose.foundation.interaction.collectIsPressedAsState
import androidx.compose.foundation.layout.defaultMinSize
import androidx.compose.foundation.rememberScrollState
import androidx.compose.foundation.selection.selectableGroup
import androidx.compose.foundation.shape.CornerSize
import androidx.compose.material.icons.Icons
import androidx.compose.material.icons.filled.Check
import androidx.compose.material.icons.outlined.Info
import androidx.compose.material3.FilledTonalButton
import androidx.compose.runtime.getValue
import androidx.compose.runtime.remember
import androidx.compose.ui.semantics.Role
import androidx.compose.ui.semantics.role
import androidx.compose.ui.semantics.semantics
import androidx.compose.ui.unit.sp
import com.webyar.operator.ui.design.ExpressiveShapes
import com.webyar.operator.ui.design.Motion
import com.webyar.operator.ui.design.WebyarType

/**
 * The one prominent action on a screen.
 *
 * Full width, 56dp — Material 3 Expressive's medium button — and a pill
 * that squares up under the finger, the Expressive press feedback. It swaps
 * its label for the loading indicator while busy **without changing size**:
 * a button that shrinks under the thumb mid-tap is how a mis-tap happens.
 */
@Composable
fun PrimaryButton(
    label: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
    busy: Boolean = false,
) {
    val interaction = remember { MutableInteractionSource() }
    Button(
        onClick = onClick,
        enabled = enabled && !busy,
        shape = rememberPressShape(interaction),
        interactionSource = interaction,
        contentPadding = PaddingValues(horizontal = Space.xl, vertical = Space.md),
        modifier = modifier
            .fillMaxWidth()
            // heightIn, not height: at a large font scale the label needs room
            // to grow, and a fixed height is how a button clips its own text
            // on exactly the devices where that matters most.
            .heightIn(min = Size.buttonHeight),
    ) {
        if (busy) {
            LoadingIndicator(
                color = MaterialTheme.colorScheme.onSurface.copy(alpha = 0.6f),
                size = 28.dp,
            )
        } else {
            Text(label, style = WebyarType.labelLargeEmphasized.copy(fontSize = 16.sp))
        }
    }
}

/**
 * The quieter second action, where a screen has one — tonal, so it is plainly
 * a button without competing with the primary one.
 */
@Composable
fun SecondaryButton(
    label: String,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    enabled: Boolean = true,
) {
    val interaction = remember { MutableInteractionSource() }
    FilledTonalButton(
        onClick = onClick,
        enabled = enabled,
        shape = rememberPressShape(interaction),
        interactionSource = interaction,
        contentPadding = PaddingValues(horizontal = Space.xl, vertical = Space.sm),
        modifier = modifier.heightIn(min = Size.minTouchTarget),
    ) {
        Text(label, style = MaterialTheme.typography.labelLarge)
    }
}

/**
 * One choice out of a few — Material 3 Expressive's connected button group.
 *
 * Each option is its own button with a small gap between them. The selected
 * one is filled and fully round, the rest are tonal with softer corners, and
 * the shape morphs as the selection moves: the change is seen, not just
 * read. A count rides in the label when there is one, because the whole
 * reason to glance at this control is to see where the work is.
 *
 * The options come from the plan, not from an enum's own cases — a choice
 * that leads to a permanently empty list because the plan excludes it reads
 * as a broken app, not as an upsell.
 */
@Composable
fun <T> FilterPicker(
    options: List<T>,
    selected: T,
    label: (T) -> String,
    count: (T) -> Int?,
    language: Language,
    onSelect: (T) -> Unit,
    modifier: Modifier = Modifier,
) {
    Row(
        modifier
            .fillMaxWidth()
            .horizontalScroll(rememberScrollState())
            .selectableGroup(),
        horizontalArrangement = Arrangement.spacedBy(Space.xs),
    ) {
        options.forEach { option ->
            ChoiceButton(
                label = label(option),
                count = count(option),
                selected = option == selected,
                language = language,
                onClick = { onSelect(option) },
            )
        }
    }
}

/** One button of a [FilterPicker]. */
@Composable
fun ChoiceButton(
    label: String,
    selected: Boolean,
    language: Language,
    onClick: () -> Unit,
    modifier: Modifier = Modifier,
    count: Int? = null,
) {
    val interaction = remember { MutableInteractionSource() }
    val pressed by interaction.collectIsPressedAsState()
    // Selected: a full pill. Otherwise a softer rectangle, and squarer still
    // while pressed — the press "squish" the Expressive buttons have.
    val percent by animateFloatAsState(
        targetValue = when {
            pressed -> 18f
            selected -> 50f
            else -> 30f
        },
        animationSpec = Motion.fastSpatial(),
        label = "choiceShape",
    )
    val container by animateColorAsState(
        if (selected) MaterialTheme.colorScheme.primary else MaterialTheme.colorScheme.surfaceContainerHigh,
        Motion.effects(),
        label = "choiceContainer",
    )
    val content by animateColorAsState(
        if (selected) MaterialTheme.colorScheme.onPrimary else MaterialTheme.colorScheme.onSurfaceVariant,
        Motion.effects(),
        label = "choiceContent",
    )
    Surface(
        selected = selected,
        onClick = onClick,
        shape = percentShape(percent),
        color = container,
        contentColor = content,
        interactionSource = interaction,
        modifier = modifier
            .heightIn(min = 40.dp)
            .semantics { role = Role.RadioButton },
    ) {
        Row(
            Modifier.padding(horizontal = Space.lg, vertical = Space.sm),
            verticalAlignment = Alignment.CenterVertically,
            // Centred, for when the button is stretched across a row. The
            // tick brings its own gap in with it: a gap left standing beside
            // an invisible tick would push the label off-centre.
            horizontalArrangement = Arrangement.Center,
        ) {
            AnimatedVisibility(visible = selected) {
                Row {
                    Icon(Icons.Filled.Check, contentDescription = null, modifier = Modifier.size(18.dp))
                    Spacer(Modifier.size(Space.sm))
                }
            }
            Text(
                label,
                style = if (selected) WebyarType.labelLargeEmphasized else MaterialTheme.typography.labelLarge,
                maxLines = 1,
            )
            if (count != null && count > 0) {
                Spacer(Modifier.size(Space.sm))
                Text(
                    Format.number(count, language),
                    style = MaterialTheme.typography.labelMedium,
                    color = content.copy(alpha = 0.8f),
                    maxLines = 1,
                )
            }
        }
    }
}

/**
 * Shown when a list legitimately has nothing in it.
 *
 * The icon sits on one of the Expressive shapes, which is the difference
 * between "nothing here" said by a designed product and said by a blank
 * screen. Centred as a block, with the body centre-aligned and width-limited
 * so it never runs edge to edge — a paragraph the full width of a tablet is a
 * wall, not a sentence.
 */
@Composable
fun EmptyState(
    icon: ImageVector,
    title: String,
    body: String?,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier
            .fillMaxWidth()
            .padding(horizontal = Space.xl, vertical = Space.xxl),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(Space.md),
    ) {
        ShapeFrame(
            polygon = ExpressiveShapes.softBurst,
            color = MaterialTheme.colorScheme.secondaryContainer,
            modifier = Modifier.size(104.dp),
        ) {
            Icon(
                icon,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.onSecondaryContainer,
                modifier = Modifier.size(36.dp),
            )
        }
        Text(
            title,
            style = WebyarType.titleLargeEmphasized,
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(top = Space.sm),
        )
        if (body != null) {
            Text(
                body,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center,
                modifier = Modifier.widthIn(max = 320.dp),
            )
        }
    }
}

/** Shown when a request failed and retrying is the sensible next step. */
@Composable
fun ErrorState(
    title: String,
    body: String?,
    retryLabel: String,
    onRetry: () -> Unit,
    modifier: Modifier = Modifier,
) {
    Column(
        modifier
            .fillMaxWidth()
            .padding(horizontal = Space.xl, vertical = Space.xxl),
        horizontalAlignment = Alignment.CenterHorizontally,
        verticalArrangement = Arrangement.spacedBy(Space.md),
    ) {
        ShapeFrame(
            polygon = ExpressiveShapes.cookie4,
            color = MaterialTheme.colorScheme.errorContainer,
            modifier = Modifier.size(96.dp),
        ) {
            Icon(
                Icons.Outlined.Info,
                contentDescription = null,
                tint = MaterialTheme.colorScheme.onErrorContainer,
                modifier = Modifier.size(34.dp),
            )
        }
        Text(
            title,
            style = WebyarType.titleLargeEmphasized,
            textAlign = TextAlign.Center,
            modifier = Modifier.padding(top = Space.sm),
        )
        if (body != null) {
            Text(
                body,
                style = MaterialTheme.typography.bodyMedium,
                color = MaterialTheme.colorScheme.onSurfaceVariant,
                textAlign = TextAlign.Center,
                modifier = Modifier.widthIn(max = 320.dp),
            )
        }
        SecondaryButton(retryLabel, onRetry, Modifier.padding(top = Space.sm))
    }
}

/**
 * The unread counter on an inbox row.
 *
 * Digits follow the interface language: a Persian operator reads ۱۲۸, and a
 * badge is the one place a wrong digit set is unmissable.
 */
@Composable
fun UnreadBadge(count: Int, language: Language, modifier: Modifier = Modifier) {
    if (count <= 0) return
    Surface(
        color = MaterialTheme.colorScheme.primary,
        contentColor = MaterialTheme.colorScheme.onPrimary,
        shape = RoundedCornerShape(Radius.pill),
        modifier = modifier.defaultMinSize(minWidth = 22.dp, minHeight = 22.dp),
    ) {
        Box(contentAlignment = Alignment.Center) {
            Text(
                text = Format.number(count, language),
                style = WebyarType.labelLargeEmphasized.copy(fontSize = 12.sp, lineHeight = 16.sp),
                maxLines = 1,
                modifier = Modifier.padding(horizontal = 6.dp, vertical = 3.dp),
            )
        }
    }
}

/** A small status pill — "resolved", "AI", and similar. */
@Composable
fun StatusPill(
    label: String,
    modifier: Modifier = Modifier,
    tone: PillTone = PillTone.NEUTRAL,
) {
    val container = when (tone) {
        PillTone.NEUTRAL -> MaterialTheme.colorScheme.surfaceContainerHighest
        PillTone.SUCCESS -> WebyarTheme.colors.success.copy(alpha = 0.16f)
        PillTone.WARNING -> WebyarTheme.colors.warning.copy(alpha = 0.16f)
        PillTone.BRAND -> MaterialTheme.colorScheme.primaryContainer
    }
    val content = when (tone) {
        PillTone.NEUTRAL -> MaterialTheme.colorScheme.onSurfaceVariant
        PillTone.SUCCESS -> WebyarTheme.colors.success
        PillTone.WARNING -> WebyarTheme.colors.warning
        PillTone.BRAND -> MaterialTheme.colorScheme.onPrimaryContainer
    }
    Surface(
        color = container,
        contentColor = content,
        shape = RoundedCornerShape(Radius.pill),
        modifier = modifier,
    ) {
        Text(
            label,
            style = MaterialTheme.typography.labelSmall,
            maxLines = 1,
            modifier = Modifier.padding(horizontal = Space.sm, vertical = Space.xxs),
        )
    }
}

enum class PillTone { NEUTRAL, SUCCESS, WARNING, BRAND }

/**
 * A run of text that is Latin whatever the interface language is — an address,
 * a phone number, a version string.
 *
 * Reading `operator@webyar.app` right-to-left puts the domain first, which is
 * wrong in Persian and Turkish just as it would be in English. This sets the
 * text's own direction rather than the layout's, so the string reads correctly
 * while still sitting on the right edge of a Persian row where it belongs.
 */
@Composable
fun LatinText(
    text: String,
    modifier: Modifier = Modifier,
    style: TextStyle = MaterialTheme.typography.bodyMedium,
    color: androidx.compose.ui.graphics.Color = androidx.compose.ui.graphics.Color.Unspecified,
    maxLines: Int = Int.MAX_VALUE,
    /**
     * Where the run sits in the space it is given.
     *
     * Separate from the direction, and absolute: `End` inside a forced-LTR
     * style would resolve to the right even in Persian, which is the wrong
     * edge for a value column in an RTL screen.
     */
    align: TextAlign? = null,
) {
    Text(
        text = text,
        modifier = modifier,
        style = style.copy(textDirection = TextDirection.Ltr),
        color = color,
        maxLines = maxLines,
        textAlign = align ?: TextAlign.Unspecified,
        overflow = TextOverflow.Ellipsis,
    )
}

/**
 * Reading order from the TEXT, alignment from the LAYOUT.
 *
 * [LatinText] forces one direction, which is right for an address or a version
 * string and wrong for anything that could arrive in any language — a message
 * preview, a contact's name, a subject line.
 *
 * Compose resolves an unspecified `textDirection` from the layout, so a
 * Turkish sentence inside a Persian list is laid out right-to-left and the
 * bidi algorithm moves its trailing punctuation to the far end: «Tabii, hemen
 * kontrol ediyorum.» came out as «.Tabii, hemen kontrol ediyorum», with the
 * full stop leading. `Content` resolves from the first strong character
 * instead, which is what every other platform does by default.
 *
 * The alignment has to be pinned separately, and absolutely. `TextAlign.Start`
 * resolves against the text's OWN direction once that is content-derived, so a
 * Latin row would jump to the left edge of a Persian list while its Persian
 * neighbours stayed right. Left/Right do not resolve, so the column stays
 * straight whatever language lands in it.
 */
@Composable
fun TextStyle.bidiContent(): TextStyle = copy(
    textDirection = TextDirection.Content,
    textAlign = rowTextAlign(),
)

/**
 * The row's own leading edge, stated absolutely.
 *
 * `TextAlign.Start` is not the same thing: it resolves against the text's
 * direction, so the moment a run is forced or derived to LTR inside a Persian
 * screen it jumps to the left while its neighbours stay right. Left and Right
 * do not resolve, so a column of mixed-language values stays a column.
 */
@Composable
fun rowTextAlign(): TextAlign = when (LocalLayoutDirection.current) {
    LayoutDirection.Rtl -> TextAlign.Right
    LayoutDirection.Ltr -> TextAlign.Left
}

/**
 * A centred, quiet line inside a list — "nothing here yet", said without
 * making a scene of it.
 *
 * Not an [EmptyState]: that one owns a whole screen with an icon and a title.
 * This is one row saying one section is empty while the list carries on around
 * it.
 */
@Composable
fun QuietRow(text: String, modifier: Modifier = Modifier) {
    Box(
        modifier
            .fillMaxWidth()
            .padding(horizontal = Space.screenInset, vertical = Space.lg),
        contentAlignment = Alignment.Center,
    ) {
        Text(
            text,
            style = MaterialTheme.typography.bodyMedium,
            color = WebyarTheme.colors.labelTertiary,
            textAlign = TextAlign.Center,
        )
    }
}

/** A hairline, at the inset the rows use rather than edge to edge. */
@Composable
fun RowDivider(modifier: Modifier = Modifier, inset: Boolean = true) {
    androidx.compose.material3.HorizontalDivider(
        modifier = modifier.padding(start = if (inset) Space.screenInset else 0.dp),
        thickness = Size.hairline,
        color = MaterialTheme.colorScheme.outlineVariant,
    )
}
